src/TolerantDefinitionResolver.php

```diff
@@ -0,0 +1,1177 @@
+            // ?:
+            if ($expr->ifExpression === null) {
+                return new Types\Compound([
+                    $this->resolveExpressionNodeToType($expr->condition), // TODO: why?
+                    $this->resolveExpressionNodeToType($expr->elseExpression)
+                ]);
+            }
+            // Ternary is a compound of the two possible values
+            return new Types\Compound([
+                $this->resolveExpressionNodeToType($expr->ifExpression),
+                $this->resolveExpressionNodeToType($expr->elseExpression)
+            ]);
+        }
+
+        // NULL COALLESCE
+        //   $rightOperand ?? $leftOperand => resolves to type of $rightOperand or $leftOperand
+        if ($expr instanceof Tolerant\Node\Expression\BinaryExpression && $expr->operator->kind === Tolerant\TokenKind::QuestionQuestionToken) {
+            // ?? operator
+            return new Types\Compound([
+                $this->resolveExpressionNodeToType($expr->leftOperand),
+                $this->resolveExpressionNodeToType($expr->rightOperand)
+            ]);
+        }
+
+        // BOOLEAN EXPRESSIONS: resolve to Types\Boolean
+        //   (bool) $expression
+        //   !$expression
+        //   empty($var)
+        //   isset($var)
+        //   >, >=, <, <=, &&, ||, AND, OR, XOR, ==, ===, !=, !==
+        if (
+            TolerantParserHelpers::isBooleanExpression($expr)
+
+            || ($expr instanceof Tolerant\Node\Expression\CastExpression && $expr->castType->kind === Tolerant\TokenKind::BoolCastToken)
+            || ($expr instanceof Tolerant\Node\Expression\UnaryOpExpression && $expr->operator->kind === Tolerant\TokenKind::ExclamationToken)
+            || $expr instanceof Tolerant\Node\Expression\EmptyIntrinsicExpression
+            || $expr instanceof Tolerant\Node\Expression\IssetIntrinsicExpression
+        ) {
+            return new Types\Boolean;
+        }
+
+        // STRING EXPRESSIONS: resolve to Types\String
+        //   [concatenation] .=, .
+        //   [literals] "hello", \b"hello", \B"hello", 'hello', \b'hello', HEREDOC, NOWDOC
+        //   [cast] (string) "hello"
+        //
+        //   TODO: Magic constants (__CLASS__, __DIR__, __FUNCTION__, __METHOD__, __NAMESPACE__, __TRAIT__, __FILE__)
+        if (
+            ($expr instanceof Tolerant\Node\Expression\BinaryExpression &&
+                ($expr->operator->kind === Tolerant\TokenKind::DotToken || $expr->operator->kind === Tolerant\TokenKind::DotEqualsToken)) ||
+            $expr instanceof Tolerant\Node\StringLiteral ||
+            ($expr instanceof Tolerant\Node\Expression\CastExpression && $expr->castType->kind === Tolerant\TokenKind::StringCastToken)
+        ) {
+            return new Types\String_;
+        }
+
+        // BINARY EXPRESSIONS: 
+        // Resolve to Types\Integer if both left and right operands are integer types, otherwise Types\Float
+        //   [operator] +, -, *, **
```

Comment:
This is handling binary expressions, e.g. `1+2`, so I don't think it's the right place for that.